typescript/cli/src/utils/input.ts

```diff
@@ -53,3 +71,393 @@ export async function inputWithInfo({
+): item is NormalizedChoice<Value> {
+  return !Separator.isSeparator(item) && !item.disabled;
+}
+
+function isChecked<Value>(item: Item<Value>): item is NormalizedChoice<Value> {
+  return isSelectable(item) && Boolean(item.checked);
+}
+
+function toggle<Value>(item: Item<Value>): Item<Value> {
+  return isSelectable(item) ? { ...item, checked: !item.checked } : item;
+}
+
+function normalizeChoices<Value>(
+  choices: ReadonlyArray<SearchableCheckboxChoice<Value>>,
+): NormalizedChoice<Value>[] {
+  return choices.map((choice) => {
+    const name = choice.name ?? String(choice.value);
+    return {
+      value: choice.value,
+      name,
+      short: choice.short ?? name,
+      description: choice.description,
+      disabled: choice.disabled ?? false,
+      checked: choice.checked ?? false,
+    };
+  });
+}
+
+function sortNormalizedItems<Value>(
+  a: NormalizedChoice<Value>,
+  b: NormalizedChoice<Value>,
+): number {
+  if (a.name > b.name) {
+    return 1;
+  } else {
+    return -1;
+  }
+}
+
+function organizeItems<Value>(
+  items: Array<Item<Value>>,
+  selectableOptionsSeparator?: Separator,
+): Array<Item<Value> | Separator> {
+  const orderedItems = [];
+
+  const checkedItems = items.filter(
+    (item) => !Separator.isSeparator(item) && item.checked,
+  ) as NormalizedChoice<Value>[];
+
+  if (checkedItems.length !== 0) {
+    orderedItems.push(new Separator('--Selected Options--'));
+
+    orderedItems.push(...checkedItems.sort(sortNormalizedItems));
+  }
+
+  orderedItems.push(
+    selectableOptionsSeparator ?? new Separator('--Available Options--'),
+  );
+
+  const nonCheckedItems = items.filter(
+    (item) => !Separator.isSeparator(item) && !item.checked,
+  ) as NormalizedChoice<Value>[];
+
+  orderedItems.push(...nonCheckedItems.sort(sortNormalizedItems));
+
+  if (orderedItems.length === 1) {
+    return [];
+  }
+
+  return orderedItems;
+}
+
+// the isUpKey function from the inquirer package is not used
+// because it detects k and p as custom keybindings that cause
+// the option selection to go up instead of writing the letters
+// in the search string
+function isUpKey(key: KeypressEvent): boolean {
+  return key.name === 'up';
+}
+
+// the isDownKey function from the inquirer package is not used
+// because it detects j and n as custom keybindings that cause
+// the option selection to go down instead of writing the letters
+// in the search string
+function isDownKey(key: KeypressEvent): boolean {
+  return key.name === 'down';
+}
+
+export const searchableCheckBox = createPrompt(
+  <Value>(
+    config: SearchableCheckboxConfig<Value>,
+    done: (value: Array<Value>) => void,
+  ) => {
+    const {
+      instructions,
+      pageSize = 7,
+      loop = true,
+      required,
+      validate = () => true,
+      selectableOptionsSeparator,
+    } = config;
+    const theme = makeTheme<SearchableCheckboxTheme>(
+      checkboxTheme,
+      config.theme,
+    );
+    const firstRender = useRef(true);
```

Comment:
So strange to see these react hooks here. I wondered if inquirer was using react under the hood but actually they implemented these themselves! https://github.com/SBoudrias/Inquirer.js/blob/main/packages/core/src/lib/use-memo.ts